tests/0147-consumer_group_consumer_mock.c

```diff
@@ -0,0 +1,682 @@
+        TEST_CALL_ERR__(rd_kafka_subscribe(c, subscription));
+        rd_kafka_topic_partition_list_destroy(subscription);
+
+        /*First HB is fenced and second receives assignment*/
+        expected_heartbeats = 2;
+        if (variation == 1)
+                /*First HB receives assignment*/
+                expected_heartbeats = 1;
+
+        TEST_SAY("Awaiting initial HBs\n");
+        TEST_ASSERT((found_heartbeats =
+                         wait_all_heartbeats_done(mcluster, expected_heartbeats,
+                                                  200)) == expected_heartbeats,
+                    "Expected %d heartbeats, got %d", expected_heartbeats,
+                    found_heartbeats);
+
+        expected_rebalance_cnt = 0;
+        /* variation 0: Second HB assigned */
+        if (variation == 1) {
+                expected_rebalance_cnt++;
+                rebalance_exp_event = RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS;
+
+                /* First HB assigned */
+                rkmessage = rd_kafka_consumer_poll(c, 100);
+                TEST_ASSERT(!rkmessage, "No message should be returned");
+
+                TEST_ASSERT(rebalance_cnt == expected_rebalance_cnt,
+                            "Expected %d rebalance events after assign "
+                            "callback, got %d",
+                            expected_rebalance_cnt, rebalance_cnt);
+
+                TEST_SAY("Awaiting partition lost callback\n");
+                /* Second HB acks and loses partitions */
+                expected_rebalance_cnt++;
+                rebalance_exp_event = RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS;
+                rebalance_exp_lost  = rd_true;
+                /* Needs to wait HB interval */
+                rkmessage = rd_kafka_consumer_poll(c, 750);
+                TEST_ASSERT(!rkmessage, "No message should be returned");
+
+                TEST_ASSERT(
+                    rebalance_cnt == expected_rebalance_cnt,
+                    "Expected %d rebalance events after lost callback, got %d",
+                    expected_rebalance_cnt, rebalance_cnt);
+
+                /* Third HB assigns again */
+        }
+
+        expected_rebalance_cnt++;
+        rebalance_exp_event = RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS;
+        rebalance_exp_lost  = rd_false;
+
+        TEST_SAY("Clearing mock requests\n");
+        rd_kafka_mock_clear_requests(mcluster);
+        expected_heartbeats = 0;
+
+        TEST_SAY("Awaiting rebalance callback\n");
+        /* Consume from c, partitions are lost if assigned */
+        rkmessage = rd_kafka_consumer_poll(c, 500);
+        TEST_ASSERT(!rkmessage, "No message should be returned");
+
+        TEST_ASSERT(rebalance_cnt == expected_rebalance_cnt,
+                    "Expected %d total rebalance events, got %d",
+                    expected_rebalance_cnt, rebalance_cnt);
+
+        if (variation == 0) {
+                /* Ack for assignment HB */
+                expected_heartbeats++;
```

Comment:
I want to understand how can we count the number of heartbeats? Its periodic with 1s time period. Is our assumption that we will execute all these operations within 1s so that we only consider expedited heartbeats?

Reply:
When it happens on first HB it should be as fast as possible after the fencing. On variation 1 it's fast too as the ACK should come immediately on second HB if the callback is processed soon, on second HB there's the ACK and the fencing.